### PR TITLE
In-memory authentication cache

### PR DESCRIPTION
**Pull request by lolski**
## Release notes: product changes
Improve authentication verification performance by introducing an in-memory cache. The cached credential will be invalidated after some specified time period and also if either the password is updated or the user account itself is deleted.

## Motivation
Improve the speed at which authentication verification is performed.

## Implementation
Implement authentication cache with Moka's cache, that behaves as follows:
- cache user password in memory if the authenticator successfully verifies the credential against the one stored in the database
- invalidate the cache when the password is updated, or if the user is deleted
- a cache entry is invalidated after a certain time:
  - every 3 hours
  - alternatively, after 1 hour of idle time (ie., no new request are received from the user)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -305,7 +305,7 @@ workflows:
       - sync-to-typedb-benchmark:
           filters:
             branches:
-              only: [ "3.0" ]
+              only: [ "3.0", "authentication-cache" ]
```
Comment (lolski) [inline .circleci/config.yml:308]: TODO: remove before merging

File: server/authenticator.rs [hunk 2]
```diff
@@ -37,15 +40,28 @@ impl Authenticator {
...
-            return Err(Status::unauthenticated(ERROR_INVALID_CREDENTIAL));
-        };
-
-        if password_hash.matches(password) {
-            Ok(http::Request::from_parts(parts, body))
-        } else {
-            Err(Status::unauthenticated(ERROR_INVALID_CREDENTIAL))
-        }
+        Ok(http::Request::from_parts(parts, body))
+        // match self.authenticator_cache.get_user(username) {
```
Comment (flyingsilverfin) [inline server/authenticator.rs:44]: uncomment right?

File: server/typedb.rs [hunk 3]
```diff
@@ -56,10 +59,10 @@ impl Server {
 
     pub async fn serve(mut self) -> Result<(), tonic::transport::Error> {
         let service = typedb_protocol::type_db_server::TypeDbServer::new(self.typedb_service.take().unwrap());
+        // let authenticator = Authenticator::new(self.user_manager.clone(), self.authenticator_cache.clone());
```
Comment (flyingsilverfin) [inline server/typedb.rs:62]: commented stuff

File: user/user_manager.rs [hunk 4]
```diff
@@ -99,7 +99,10 @@ impl UserManager {
             },
         );
         match update_result {
-            Ok(Ok(())) => Ok(()),
+            Ok(Ok(())) => {
+                // TODO: invalidate cache
```
Comment (flyingsilverfin) [inline user/user_manager.rs:103]: implement

Comment (lolski) [reply user/user_manager.rs]: This is an outdated todo, and has already been [implemented in `TypeDBService`](https://github.com/typedb/typedb/pull/7287/files?show-viewed-files=true&file-filters%5B%5D=#diff-86c12c24889b5c2ffc1b18bb1ff46d8c7a137d416d8dd24f14cda7a87180523cR239)

File: user/user_manager.rs [hunk 5]
```diff
@@ -135,7 +138,10 @@ impl UserManager {
             },
         );
         match delete_result {
-            Ok(Ok(())) => Ok(()),
+            Ok(Ok(())) => {
+                // TODO: invalidate cache
```
[the remark this answers is not in the document]
Comment (lolski) [reply user/user_manager.rs]: This is an outdated todo, and has already been [implemented in `TypeDBService`](https://github.com/typedb/typedb/pull/7287/files?show-viewed-files=true&file-filters%5B%5D=#diff-86c12c24889b5c2ffc1b18bb1ff46d8c7a137d416d8dd24f14cda7a87180523cR262)

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**